elements/lisk-tree/src/sparse_merkle_tree/utils.ts

```diff
@@ -13,9 +13,180 @@
+export const filterQueries = <T extends InclusionProofQuery>(
+	queries: T[],
+	keyLength: number,
+): T[] => {
+	const uniqueKeys: string[] = [];
+
+	return queries.filter(q => {
+		const binaryBitmap = bufferToBinaryString(q.bitmap);
+		const h = binaryBitmap.length;
+		const binaryKey = binaryExpansion(q.key, keyLength);
+		const keyPrefix = binaryKey.substring(0, h);
+
+		if (!uniqueKeys.includes(keyPrefix)) {
+			uniqueKeys.push(keyPrefix);
+			return true;
+		}
+
+		return false;
+	});
+};
+
+// Checks whether two queries correspond to nodes that are
+// children of the same branch node, with q1 and q2 the left and right
+// child respectively
+// https://github.com/LiskHQ/lips-staging/blob/master/proposals/lip-0039.md#proof-verification
+export const areSiblingQueries = (
+	q1: InclusionProofQuery,
+	q2: InclusionProofQuery,
+	keyLength: number,
+): boolean => {
+	const q1BinaryBitmap = bufferToBinaryString(q1.bitmap);
+	const q2BinaryBitmap = bufferToBinaryString(q2.bitmap);
+
+	if (q1BinaryBitmap.length !== q2BinaryBitmap.length) {
+		return false;
+	}
+
+	const h = q1BinaryBitmap.length;
+	const binaryKey1 = binaryExpansion(q1.key, keyLength);
+	const binaryKey2 = binaryExpansion(q2.key, keyLength);
+
+	// end of string is exclusive
+	const keyPrefix1 = binaryKey1.substring(0, h - 1);
+	const keyPrefix2 = binaryKey2.substring(0, h - 1);
+
+	if (keyPrefix1 !== keyPrefix2) {
+		return false;
+	}
+
+	const d1 = binaryKey1[h];
+	const d2 = binaryKey2[h];
+
+	return (d1 === '0' && d2 === '1') || (d1 === '1' && d2 === '0');
+};
+
+// Calculate root for the given sibling hashes
+// https://github.com/LiskHQ/lips-staging/blob/master/proposals/lip-0039.md#proof-verification
+export const calculateRoot = (
+	sibHashes: Buffer[],
+	queries: InclusionProofQuery[],
+	keyLength: number,
+): Buffer => {
+	let siblingHashes = objects.cloneDeep(sibHashes);
+	const data: InclusionProofQueryWithHash[] = [];
+
+	for (const q of queries) {
+		data.push({
+			...q,
+			binaryBitmap: bufferToBinaryString(q.bitmap),
+			hash: q.value.byteLength === 0 ? EMPTY_HASH : leafHash(q.key, q.value),
+		});
+	}
+
+	let sortedQueries = filterQueries<InclusionProofQueryWithHash>(
+		sortByBitmapAndKey<InclusionProofQueryWithHash>(data),
+		keyLength,
+	);
+
+	while (sortedQueries.length > 0) {
+		const q = sortedQueries[0];
+
+		// if the binaryBitmap is empty string, we reached the top of the tree
+		if (q.binaryBitmap === '') {
+			return q.hash;
+		}
+
+		const b = q.binaryBitmap[0];
+		// h equals the height of the node; e.g., the root has h=0
+		const h = q.binaryBitmap.length;
+		const binaryKey = binaryExpansion(q.key, keyLength);
+
+		// we distinguish three cases for the sibling hash:
+		// 1. sibling is next element of sortedQueries
+		let siblingHash!: Buffer;
+
+		if (sortedQueries.length > 1 && areSiblingQueries(q, sortedQueries[1], keyLength)) {
+			siblingHash = sortedQueries[1].hash;
+			sortedQueries = sortedQueries.splice(1, 1);
```

Comment:
return value of `splice` is the removed element, but here i think you just need to remove?